File: packages/a2a/src/_jwt.ts
/**
 * Lightweight offline JWT decode for A2A grant token inspection.
 *
 * For full verification with JWKS, use the @grantex/sdk tokens.verify() method.
 * This module only performs base64 decoding of the JWT payload.
 */

export interface JwtPayload {
  iss?: string;
  sub?: string;
  agt?: string;
  dev?: string;
  scp?: string[];
  jti?: string;
  grnt?: string;
  aud?: string;
  iat?: number;
  exp?: number;
  bdg?: number;
  parentAgt?: string;
  parentGrnt?: string;
  delegationDepth?: number;
}

/**
 * Decode a JWT payload without verification.
 * Throws if the token is not a valid JWT format.
 */
export function decodeJwtPayload(token: string): JwtPayload {
  const parts = token.split('.');
  if (parts.length !== 3) {
    throw new Error('Invalid JWT format: expected 3 parts');
  }

  const payload = parts[1]!;
  const decoded = Buffer.from(payload, 'base64url').toString('utf8');
  return JSON.parse(decoded) as JwtPayload;
}

/**
 * Check if a decoded JWT is expired.
 */
export function isTokenExpired(payload: JwtPayload): boolean {
  if (!payload.exp) return false;
  return Date.now() >= payload.exp * 1000;
}

File: packages/a2a-py/src/grantex_a2a/_jwt.py
"""Lightweight offline JWT decode for A2A grant token inspection."""

from __future__ import annotations

import base64
import json
import time
from typing import Any, Dict, cast


def decode_jwt_payload(token: str) -> Dict[str, Any]:
    """Decode a JWT payload without verification.

    Args:
        token: The JWT string (header.payload.signature)

    Returns:
        The decoded JWT payload as a dictionary.

    Raises:
        ValueError: If the token is not a valid JWT format.
    """
    parts = token.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid JWT format: expected 3 parts")

    payload = parts[1]
    # Add padding if needed
    padding = 4 - len(payload) % 4
    if padding != 4:
        payload += "=" * padding

    decoded = base64.urlsafe_b64decode(payload)
    decoded_payload: object = json.loads(decoded)
    if not isinstance(decoded_payload, dict):
        raise ValueError("Invalid JWT payload: expected a JSON object")
    return cast(Dict[str, Any], decoded_payload)


def is_token_expired(payload: Dict[str, Any]) -> bool:
    """Check if a decoded JWT is expired.

    Args:
        payload: The decoded JWT payload.

    Returns:
        True if the token is expired, False otherwise.
    """
    exp = payload.get("exp")
    if exp is None:
        return False
    if isinstance(exp, bool) or not isinstance(exp, (int, float)):
        raise ValueError("Invalid JWT exp claim: expected a number")
    return time.time() >= float(exp)
